### ISSUE TITLE
[Typescript] Improve sql() types

### ISSUE THREAD

**Issue by xiaoyu-tamu** [porsager/postgres#85]
Looks like that we cannot pass pre defined interface to sql() helper. @Minigugus

```ts
// table.genarated.ts

interface ExchangeRate {
  date: Date;
  source: string;
  target: string;
  rate: number;
}
export interface ExchangeRateSelect extends SetRequired<ExchangeRate> {}
export interface ExchangeRateInsert extends ExchangeRate {}
export interface ExchangeRateUpdate extends Partial<ExchangeRate> {}
```

```ts
    const data: ExchangeRateInsert[] = [];

    // not work
    const result = await sql`
      INSERT INTO exchange_rate 
      ${sql(data)}
      ON CONFLICT ON CONSTRAINT exchange_rate_pkey
      DO NOTHING`;


    // not work
    const result = await sql`
      INSERT INTO exchange_rate 
      ${sql<ExchangeRateInsert[],[]>(data)}
      ON CONFLICT ON CONSTRAINT exchange_rate_pkey
      DO NOTHING`;

    // not work
    const result = await sql`
      INSERT INTO exchange_rate 
      ${sql<ExchangeRateInsert[]>(data)}
      ON CONFLICT ON CONSTRAINT exchange_rate_pkey
      DO NOTHING`;


```

```
No overload matches this call.
  Overload 1 of 4, '(template: TemplateStringsArray, ...args: SerializableParameter[]): PendingQuery<Row[]>', gave the following error.
    Argument of type 'PendingQuery<Row[]>' is not assignable to parameter of type 'SerializableParameter'.
      Type 'PendingQuery<Row[]>' is missing the following properties from type 'SerializableParameter[]': length, pop, push, concat, and 28 more.
  Overload 2 of 4, '(...columns: string[]): Helper<string, string[]>', gave the following error.
    Argument of type 'TemplateStringsArray' is not assignable to parameter of type 'string'.
  Overload 3 of 4, '(objOrArray: HelperSerializable, ...keys: JobId[]): Helper<HelperSerializable, JobId[]>', gave the following error.
    Argument of type 'TemplateStringsArray' is not assignable to parameter of type 'HelperSerializable'.
      Type 'TemplateStringsArray' is not assignable to type '{ [index: string]: SerializableParameter; }'.
        Index signature is missing in type 'TemplateStringsArray'.ts(2769)
No overload matches this call.
  The last overload gave the following error.
    Argument of type 'ExchangeRateInsert[]' is not assignable to parameter of type 'HelperSerializable'.
      Type 'ExchangeRateInsert[]' is not assignable to type '{ [index: string]: SerializableParameter; }[]'.
        Type 'ExchangeRateInsert' is not assignable to type '{ [index: string]: SerializableParameter; }'.
          Index signature is missing in type 'ExchangeRateInsert'.ts(2769)
index.d.ts(312, 5): The last overload is declared here.

```

**Updated**

However, it works if you define these type inline or cast it. 

```ts
const result = await sql`
  INSERT INTO exchange_rate 
  ${sql<{ 
       date: Date;
       source: string;
       target: string;
       rate: number;
   }[],[]>(data)}
  ON CONFLICT ON CONSTRAINT exchange_rate_pkey
  DO NOTHING`;


const result = await sql`
  INSERT INTO exchange_rate 
  ${sql<Readonly<ExchangeRateInsert>[],[]>(data)}
  ON CONFLICT ON CONSTRAINT exchange_rate_pkey
  DO NOTHING`;
  
```

**Comment by Minigugus** [comment]
TLDR workaround: `npm remove @types/node` :laughing: (no joke, it really solve this issue for now!)

In order to keep `postgres` with 0 dependencies, I prefered not to install `@types/node` as dev dependencies despite using [`Buffer` here](https://github.com/porsager/postgres/blob/c55304eb1a5c107b60c6b46e9469dfd5db1cc534/types/index.d.ts#L233). But I also forgot to install `@types/node` locally so `Buffer` was infered as `any` and my tests passed correctly, so I missed `Index signature is missing in type`... :unamused: 

> However, it works if you define these type inline or cast it.

Ok so I'm not the only guilty :sweat_smile:  https://github.com/microsoft/TypeScript/issues/15300
If you really want your `interface` you can wrap your type in `type Noop<T> = { [k in keyof T]: T[k]; };`...

I tried to get the strongest types I could to prevent unexpected SQL errors at runtime but sometimes they are too strong :laughing: 

I'm working on a fix, [what about this one](https://www.staging-typescript.org/play?#code/CYUwxgNghgTiAEYD2A7AzgF3mgjhAXPAMoCKAMgNwBQVAlihiDAGZRgICiAHmABZQoA5iABKURvADeVePGDiQhACILqstEgCuMdoUwx6gtfAyxhGPRgNDjMBYRSaAtgCMm1AL5UQXAA5IYLHpGFjZOHn4hUQUiEAhwLB9GFGA0eBEQHE1aOGAAHm4+AWExRgA+KS8ff0D4YKZWdnhCyJKFAEl0JkSuZNTmiOLoiUkqvwCghgawgaKo0pAAVV95CSSQFLSABVgMWigIAsH5hQrRmlwIAAokAGsASmpLm9vFlFpUR6pnu4BBGDsAE8vs8UEgMAB5W7-IEwqDA6gAekR8AAAhg0ABaaoJbEAgLfPBXMGQ25bAxOWh7ABuIC+hOudwANPAAOSrECskFEu5vD4oFnshRcpEo9FYnFgDB4mAEn6vd6oQUc1mCjTadgihkvOGA5XC7nXElQ3W6lkARi+yLRGOxflxTFlMG1xrJFKptFpgupB00IAhzC1VGt4rtvgd+Od8sFYJQnKtYttkuljrlPIV-JjqHjoptEvtUplacZ0IB8Kzca1IaTBZTkZd4JNZcBZrZsZzwcT+fDhdTUaJrvJtEpNJA3t9-sD9PqoSaIiQAHcpDJ4ABtZAQZwoSzWQQAXUIAkBnhoGEB4fgAAk4uGYLEDAdaAAvKAueJ5AAq8HWm3g5KQt5ngA0iAerwIs369Bs-T-oBgIgYCFQALzwHGtIwJBfRpIsVAAPzpIu8AAD4EQuq57lQhBfj+-SOK4TB4VIa70KAXB1Cg4EHvASAuAAVgk8AeORlFMauLE+OxnGEDx-FSoJ1B0FMs4INeEC3p+LIQTRaRHuR8AoR+5FnCuzA5JgVG2CA5ngSeikhI0CCkGQy6yJ+mHQWkqm3ve+wQM+r7vrcoFIMw8Afppq50W4MB7mUmnub+VxBYCIXwFc1FQb+ul7vA+GGVFTA5VR9z3EZNx8RCMC6lRLIAHT1claCEIs9yEF5TAaeBZS2aAkCwAgyDoFgdyEC0QwLLE8RStQvXQHAiCoJg3EZqgo3HG0jCTQJJGSHI9ioc40XyVQs39QtQ3LdVsytMMIBbVK5EzeAc0DYtWCuqazZrXMG13XECTkY9J3PWdg1LYO7qjg4h3uFQQA)? *(`@ts-expect-error` indicate the line should fail)*

**Comment by porsager** [comment]
@Minigugus can this be closed after merging #84 ?

**Comment by Minigugus** [comment]
@porsager The #84 didn't fixed this since it wasn't clear whether the integration of Node types should have been conserved or dropped ([it seems that `postgres` accepts `Buffer` but not `Unit8Array`, even if `Buffer` extends `Uint8Array`](https://github.com/porsager/postgres/blob/995bc7eba9b0e6aa3a75420821a177fee30c7834/lib/types.js#L100)).

Do you prefer to drop Node types integration and supports `Uint8Array` instead of `Buffer`, or to fix Node types integration by adding `@types/node` dev dependency to the project (which would break the 0 dependency principle)?

**Comment by porsager** [comment]
Ah, we should definitely just support Uint8Array :)

**Comment by porsager** [comment]
Hmm. Switching to Uint8Array is not an option since we can't do allocUnsafe. 

What's the alternative to fix this issue?

**Comment by Minigugus** [comment]
> Switching to Uint8Array is not an option since we can't do allocUnsafe.

Why couldn't we do `allocUnsafe`? As explained in the NodeJS docs, [`Buffer` extends `Uint8Array`](https://nodejs.org/api/buffer.html#buffer_buffers_and_typedarrays) and [`allocUnsafe` is static](https://nodejs.org/api/buffer.html#buffer_static_method_buffer_allocunsafe_size), so `const buffer: Uint8Array = Buffer.allocUnsafe(42);` in TypeScript should just work (downcast). However, if you need `Buffer` methods from a user provided `Uint8Array` buffer (or any *TypedArray* buffer), you can still use [Buffer.from(arrayBuffer[, byteOffset[, length]])](https://nodejs.org/api/buffer.html#buffer_static_method_buffer_from_arraybuffer_byteoffset_length) to create a new `Buffer` instance that share the user-provided buffer memory (no copy).

Note: only user-provided buffers matter here; there is no need to convert `Buffer` to `Uint8Array` in the internal code

**Comment by porsager** [comment]
Ah, I misunderstood then. So is this only something that needs fixing in the types?

**Comment by Minigugus** [comment]
> Ah, I misunderstood then. So is this only something that needs fixing in the types?

Yes, but that may also break the internal code if buffers are expected to be extends `Buffer` (for instance `toString('base64')` on a tagged template parameter won't work). It's not safe to just change types in case a user pass an `Uint8Array` instance instead of `Buffer`.

**Comment by porsager** [comment]
The only place that happens is in the `bytea` serializer right?

I already wanted to send bytea (buffer) as binary in the protocol instead of a hex string to improve performance, so that could even make it go straight through as Uint8Array.

Would that be sufficient?

**Comment by porsager** [comment]
Even so, wouldn't Buffer.from(bufferOrUint8Array).tostring('hex') in the `bytea` serializer be sufficient?

**Comment by Minigugus** [comment]
```js
Buffer.from(bufferOrUint8Array.buffer, bufferOrUint8Array.byteOffset, bufferOrUint8Array.byteLength)
      .tostring('hex')
```
should work, indeed (not tested). The extra parameters ensure the new buffer share the input buffer memory instead of copying it.

**Comment by porsager** [comment]
Ah, very cool! Let's go for that now then.

**Comment by porsager** [comment]
Do you want to make a PR @Minigugus  ?
